js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx

```diff
@@ -23,99 +25,177 @@ import {repoAddressAsHumanString} from '../workspace/repoAddressAsString';
-    // skip the cache entirely.
-    fetchPolicy: 'no-cache',
-    variables: {
-      inProgressFilter: {
-        ...runsFilter,
-        statuses: [RunStatus.CANCELING, RunStatus.STARTED],
-        createdBefore: endSec,
-      },
-    },
+  const buckets = useMemo(() => {
+    const buckets = [];
+    for (let time = start; time < end; time += BUCKET_SIZE) {
+      buckets.push([time, Math.min(end, time + BUCKET_SIZE)] as const);
+    }
+    return buckets;
+  }, [start, end]);
+
+  const singleBucket = useMemo(() => [range], [range]);
+
+  const client = useApolloClient();
+
+  const [completedRunsQueryData, setCompletedRunsData] = useState<{
+    data: RunTimelineFragment[] | undefined;
+    loading: boolean;
+    error: any;
+    called: boolean;
+  }>({
+    data: undefined,
+    loading: true,
+    error: undefined,
+    called: false,
   });
+  const [ongoingRunsQueryData, setOngoingRunsData] = useState<{
+    data: RunTimelineFragment[] | undefined;
+    loading: boolean;
+    error: any;
+    called: boolean;
+  }>({
+    data: undefined,
+    loading: true,
+    error: undefined,
+    called: false,
+  });
+  const {data: ongoingRunsData, loading: loadingOngoingRunsData} = ongoingRunsQueryData;
+  const {data: completedRunsData, loading: loadingCompletedRunsData} = completedRunsQueryData;
+
+  const fetchCompletedRunsQueryData = useCallback(async () => {
+    return await fetchPaginatedBucketData({
+      buckets,
+      setQueryData: setCompletedRunsData,
+      async fetchData(bucket, cursor: string | undefined) {
+        const {data} = await client.query<
+          CompletedRunTimelineQuery,
+          CompletedRunTimelineQueryVariables
+        >({
+          query: COMPLETED_RUN_TIMELINE_QUERY,
+          notifyOnNetworkStatusChange: true,
+          fetchPolicy: 'no-cache',
+          variables: {
+            completedFilter: {
+              ...runsFilter,
+              statuses: Array.from(doneStatuses),
+              createdBefore: bucket[1] / 1000,
+              updatedAfter: bucket[0] / 1000,
+            },
+            cursor,
+            limit: BATCH_LIMIT,
+          },
+        });
+
+        if (data.completed.__typename !== 'Runs') {
+          return {
+            data: [],
+            cursor: undefined,
+            hasMore: false,
+            error: data.completed,
+          };
+        }
+        const runs = data.completed.results;
+        const hasMoreData = runs.length === BATCH_LIMIT;
+        const nextCursor = hasMoreData ? runs[runs.length - 1]!.id : undefined;
+        return {
+          data: runs,
+          cursor: nextCursor,
+          hasMore: hasMoreData,
+          error: undefined,
+        };
+      },
+    });
+  }, [buckets, client, runsFilter]);
+
+  const fetchOngoingRunsQueryData = useCallback(async () => {
+    return await fetchPaginatedBucketData({
+      buckets: singleBucket, // We don't actually bucket the ongoing runs since it doesn't matter when they start
```

Comment:
why go through this abstraction if it should not be bucketed?

Reply:
I guess I break up this abstraction into two parts

Reply:
(but originally I did think I needed it, I'll update this)